cirq/ops/phased_x_z_gate.py

```diff
@@ -0,0 +1,179 @@
+            linalg.deconstruct_single_qubit_matrix_into_angles(mat))
+        pre_phase /= np.pi
+        post_phase /= np.pi
+        rotation /= np.pi
+        pre_phase -= 0.5
+        post_phase += 0.5
+        return PhasedXZPowGate(x_exponent=rotation,
+                               axis_phase_exponent=-pre_phase,
+                               z_exponent=post_phase + pre_phase)._canonical()
+
+    def _qasm_(self, args: 'cirq.QasmArgs',
+               qubits: Tuple['cirq.Qid', ...]) -> Optional[str]:
+        from cirq.circuits import qasm_output
+        qasm_gate = qasm_output.QasmUGate(lmda=0.5 - self._axis_phase_exponent,
+                                          theta=self._x_exponent,
+                                          phi=self._z_exponent +
+                                          self._axis_phase_exponent - 0.5)
+        return protocols.qasm(qasm_gate, args=args, qubits=qubits)
+
+    def _decompose_(self, qubits: Sequence['cirq.Qid']) -> 'cirq.OP_TREE':
+        q = qubits[0]
+        yield ops.Z(q)**-self._axis_phase_exponent
+        yield ops.X(q)**self._x_exponent
+        yield ops.Z(q)**(self._z_exponent + self._axis_phase_exponent)
+
+    def __pow__(self, exponent: Union[float, int]) -> 'PhasedXZPowGate':
+        if exponent == 1:
+            return self
+        if exponent == -1:
+            return PhasedXZPowGate(
+                x_exponent=-self._x_exponent,
+                z_exponent=-self._z_exponent,
+                axis_phase_exponent=self._z_exponent + self.axis_phase_exponent,
+            )
+        return NotImplemented
+
+    def _is_parameterized_(self) -> bool:
+        """See `cirq.SupportsParameterization`."""
+        return (protocols.is_parameterized(self._x_exponent) or
+                protocols.is_parameterized(self._z_exponent) or
+                protocols.is_parameterized(self._axis_phase_exponent))
+
+    def _resolve_parameters_(self, param_resolver) -> 'cirq.PhasedXZPowGate':
+        """See `cirq.SupportsParameterization`."""
+        return PhasedXZPowGate(
+            z_exponent=param_resolver.value_of(self._z_exponent),
+            x_exponent=param_resolver.value_of(self._x_exponent),
+            axis_phase_exponent=param_resolver.value_of(
+                self._axis_phase_exponent),
+        )
+
+    def _phase_by_(self, phase_turns, qubit_index) -> 'cirq.PhasedXZPowGate':
+        """See `cirq.SupportsPhase`."""
+        assert qubit_index == 0
+        return PhasedXZPowGate(x_exponent=self._x_exponent,
+                               z_exponent=self._z_exponent,
+                               axis_phase_exponent=self._axis_phase_exponent +
+                               phase_turns * 2)
+
+    def _circuit_diagram_info_(self,
+                               args: 'cirq.CircuitDiagramInfoArgs') -> str:
+        """See `cirq.SupportsCircuitDiagramInfo`."""
+        return (f'PhXZ('
+                f'p={args.format_real(self._axis_phase_exponent)},'
```

Comment:
Call this `a` instead to match constructor args?

Reply:
Done.